### PR TITLE
Add OpenGL support for :class:`~.StreamLines`

### PR DESCRIPTION
**Pull request by ryanmccauley211**
## Overview: What does this pull request change?

Add OpenGL support for :class:~.StreamLines

## Motivation and Explanation: Why and how do your changes improve the library?

Add opengl support for StreamLines

```py
class BasicUsage(Scene):
    def construct(self):
        func = lambda pos: ((pos[0] * UR + pos[1] * LEFT) - pos) / 3
        self.add(StreamLines(func))
```

BasicUsage_ManimCE_v0 9 0

```py
class SpawningAndFlowingArea(Scene):
    def construct(self):
        func = lambda pos: np.sin(pos[0]) * UR + np.cos(pos[1]) * LEFT + pos / 5
        stream_lines = StreamLines(
            func,
            x_min=-3, x_max=3, delta_x=0.2,
            y_min=-2, y_max=2, delta_y=0.2,
            padding=1
        )

        spawning_area = Rectangle(width=6, height=4)
        flowing_area = Rectangle(width=8, height=6)
        labels = [
            Tex("Spawning Area"),
            Tex("Flowing Area").shift(DOWN * 2.5)
        ]
        for lbl in labels:
            lbl.add_background_rectangle(opacity=0.6, buff=0.05)

        self.add(stream_lines, spawning_area, flowing_area, *labels)
```

SpawningAndFlowingArea_ManimCE_v0 9 0

```py
class BasicUsage(Scene):
    def construct(self):
        func = lambda pos: ((pos[0] * UR + pos[1] * LEFT) - pos) / 3.
        self.add(StreamLines(func, colors=[BLUE, PINK, PURPLE], stroke_width=6., opacity=0.2))
        self.interactive_embed()
```

Screenshot 2021-08-30 at 09 38 00

## Further Information and Comments

TODO - figure out why the gradient ends unevenly with the tips colored. EDIT - this seems to be system / gfx card related. Above images on a different system do not have the yellow tips.

Screenshot 2021-08-29 at 11 12 26

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/mobject/vector_field.py [hunk 1]
```diff
@@ -720,20 +748,29 @@ def outside_box(p):
...
+                    ]
+                    line.set_rgba_array_direct(self.values_to_rgbas(norms, opacity))
+                else:
+                    line.color_using_background_image(self.background_img)
+                    line.set_stroke(width=self.stroke_width, opacity=opacity)
             self.add(line)
         self.stream_lines = [*self.submobjects]
 
+    def get_norm(self, vect):
+        return sum(x ** 2 for x in vect) ** 0.5
```
Comment (hydrobeam) [inline manim/mobject/vector_field.py:772]:
```suggestion
```

File: manim/mobject/vector_field.py [hunk 2]
```diff
@@ -385,6 +382,31 @@ def get_colored_background_image(self, sampling_rate: int = 5) -> Image.Image:
         rgbs = np.apply_along_axis(self.pos_to_rgb, 2, points_array)
         return Image.fromarray((rgbs * 255).astype("uint8"))
 
+    def get_vectorized_rgba_gradient_function(
```
Comment (hydrobeam) [inline manim/mobject/vector_field.py:385]: **(suggestion)**: an explanation for this method + its parameters would be helpful.

File: manim/mobject/vector_field.py [hunk 3]
```diff
@@ -720,20 +748,29 @@ def outside_box(p):
...
-                # line.set_stroke([color_func(p) for p in line.get_anchors()])
-                # TODO use color_from_background_image
-                line.color_using_background_image(self.background_img)
-            line.set_stroke(width=self.stroke_width, opacity=opacity)
+                if config["renderer"] == "opengl":
+                    # scaled for compatibility with cairo
+                    line.set_stroke(width=self.stroke_width / 4.0)
+                    norms = [
+                        self.get_norm(self.func(point)) for point in line.get_points()
+                    ]
```
Comment (hydrobeam) [inline manim/mobject/vector_field.py:763]: iirc `get_norm` was dropped in favour of numpy's `np.linalg.norm`, so it should be called directly here.

plus, `get_points` was deprecated.

```suggestion
                    norms = np.array([
                        np.linalg.norm(self.func(point)) for point in line.points
                    ])
```

File: manim/mobject/vector_field.py [hunk 4]
```diff
@@ -720,20 +748,29 @@ def outside_box(p):
             step = max_steps
             if not step:
                 continue
-            line = VMobject()
+            line = Line()
```
Comment (hydrobeam) [inline manim/mobject/vector_field.py:751]: A `Line` probably initializes some un-needed attributes, so a `VMobject` might be more appropriate. To get around incompatibility, you could use a `VGroup`, or add a check for the OpenGL renderer.

File: manim/mobject/types/opengl_vectorized_mobject.py [hunk 5]
```diff
@@ -322,6 +322,15 @@ def set_flat_stroke(self, flat_stroke=True, recurse=True):
...
 
+    def set_rgba_array_direct(
+        self,
+        rgbas: np.ndarray,
+        name="stroke_rgba",
+        recurse=True,
+    ):
+        for mob in self.get_family(recurse):
+            mob.data[name] = rgbas.copy()
+
```
Comment (hydrobeam) [inline manim/mobject/types/opengl_vectorized_mobject.py:333]: I believe this method already exists in `opengl_mobject.py`, was this just a temp change or is there a reason for it?

**Review by hydrobeam — COMMENTED** [review]
I tried out the changes, and while it does work using the default shaders written by Grant, they produce an error when using the projection shaders:

It's not a massive concern since the projection shaders are not commonly used, but we do hope to eventually switch to fully supporting them:

```py
(manim-zPbAr8g6-py3.9) PS C:\Users\laith\Documents\GitHub\hydro_manim> manim -p --renderer=opengl mkt.py BasicUsage --use_projection_stroke_shaders --use_projection_fill_shaders
Manim Community v0.9.0

Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.23.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: ╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\cli\render\commands.py:108 in render           │
│                                                                                                  │
│   105 │   │   │   while keep_running:                                                            │
│   106 │   │   │   │   for SceneClass in scene_classes_from_file(file):                           │
│   107 │   │   │   │   │   scene = SceneClass(renderer)                                           │
│ ❱ 108 │   │   │   │   │   status = scene.render()                                                │
│   109 │   │   │   │   │   if status:                                                             │
│   110 │   │   │   │   │   │   continue                                                           │
│   111 │   │   │   │   │   else:                                                                  │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\scene\scene.py:215 in render                   │
│                                                                                                  │
│    212 │   │   """                                                                               │
│    213 │   │   self.setup()                                                                      │
│    214 │   │   try:                                                                              │
│ ❱  215 │   │   │   self.construct()                                                              │
│    216 │   │   except EndSceneEarlyException:                                                    │
│    217 │   │   │   pass                                                                          │
│    218 │   │   except RerunSceneException as e:                                                  │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\mkt.py:2906 in construct                             │
│                                                                                                  │
│   2903 │   def construct(self):                                                                  │
│   2904 │   │   func = lambda pos: ((pos[0] * UR + pos[1] * LEFT) - pos) / 3.                     │
│   2905 │   │   self.add(StreamLines(func, colors=[BLUE, PINK, PURPLE], stroke_width=6., opacity  │
│ ❱ 2906 │   │   self.interactive_embed()                                                          │
│   2907                                                                                           │
│   2908 class D(Scene):                                                                           │
│   2909 │   def construct(self):                                                                  │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\scene\scene.py:1074 in interactive_embed       │
│                                                                                                  │
│   1071 │   │                                                                                     │
│   1072 │   │   self.camera.model_matrix = self.camera.default_model_matrix                       │
│   1073 │   │                                                                                     │
│ ❱ 1074 │   │   self.interact(shell, keyboard_thread)                                             │
│   1075 │                                                                                         │
│   1076 │   def interact(self, shell, keyboard_thread):                                           │
│   1077 │   │   event_handler = RerunSceneHandler(self.queue)                                     │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\scene\scene.py:1131 in interact                │
│                                                                                                  │
│   1128 │   │   │   │   self.renderer.animation_start_time = 0                                    │
│   1129 │   │   │   │   dt = time.time() - last_time                                              │
│   1130 │   │   │   │   last_time = time.time()                                                   │
│ ❱ 1131 │   │   │   │   self.renderer.render(self, dt, self.moving_mobjects)                      │
│   1132 │   │   │   │   self.update_mobjects(dt)                                                  │
│   1133 │   │   │   │   self.update_meshes(dt)                                                    │
│   1134 │   │   │   │   self.update_self(dt)                                                      │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\renderer\opengl_renderer.py:398 in render      │
│                                                                                                  │
│   395 │   │   self.window.swap_buffers()                                                         │
│   396 │                                                                                          │
│   397 │   def render(self, scene, frame_offset, moving_mobjects):                                │
│ ❱ 398 │   │   self.update_frame(scene)                                                           │
│   399 │   │                                                                                      │
│   400 │   │   if self.skip_animations:                                                           │
│   401 │   │   │   return                                                                         │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\renderer\opengl_renderer.py:418 in             │
│ update_frame                                                                                     │
│                                                                                                  │
│   415 │   │   self.refresh_perspective_uniforms(scene.camera)                                    │
│   416 │   │                                                                                      │
│   417 │   │   for mobject in scene.mobjects:                                                     │
│ ❱ 418 │   │   │   self.render_mobject(mobject)                                                   │
│   419 │   │                                                                                      │
│   420 │   │   for obj in scene.meshes:                                                           │
│   421 │   │   │   for mesh in obj.get_meshes():                                                  │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\renderer\opengl_renderer.py:302 in             │
│ render_mobject                                                                                   │
│                                                                                                  │
│   299 │   │   │   │   render_opengl_vectorized_mobject_fill(self, mobject)                       │
│   300 │   │   │                                                                                  │
│   301 │   │   │   if config["use_projection_stroke_shaders"]:                                    │
│ ❱ 302 │   │   │   │   render_opengl_vectorized_mobject_stroke(self, mobject)                     │
│   303 │   │                                                                                      │
│   304 │   │   shader_wrapper_list = mobject.get_shader_wrapper_list()                            │
│   305                                                                                            │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\renderer\vectorized_mobject_rendering.py:193   │
│ in render_opengl_vectorized_mobject_stroke                                                       │
│                                                                                                  │
│   190 │   matrix_to_mobject_list = build_matrix_lists(mobject)                                   │
│   191 │   for matrix_tuple, mobject_list in matrix_to_mobject_list.items():                      │
│   192 │   │   model_matrix = np.array(matrix_tuple).reshape((4, 4))                              │
│ ❱ 193 │   │   render_mobject_strokes_with_matrix(renderer, model_matrix, mobject_list)           │
│   194                                                                                            │
│   195                                                                                            │
│   196 def render_mobject_strokes_with_matrix(renderer, model_matrix, mobjects):                  │
│                                                                                                  │
│ C:\Users\laith\Documents\GitHub\hydro_manim\manim\renderer\vectorized_mobject_rendering.py:213   │
│ in render_mobject_strokes_with_matrix                                                            │
│                                                                                                  │
│   210 │   │   end_offset = write_offset + submob.points.shape[0]                                 │
│   211 │   │                                                                                      │
│   212 │   │   points[write_offset:end_offset] = submob.points                                    │
│ ❱ 213 │   │   colors[write_offset:end_offset] = np.repeat(                                       │
│   214 │   │   │   submob.stroke_rgba,                                                            │
│   215 │   │   │   submob.points.shape[0],                                                        │
│   216 │   │   │   axis=0,                                                                        │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValueError: could not broadcast input array from shape (133956,4) into shape (366,4)
```

File: manim/mobject/types/opengl_vectorized_mobject.py (hunk 5 above)
Comment (ryanmccauley211) [reply manim/mobject/types/opengl_vectorized_mobject.py:333]: Yes I added it in another PR, but did this before it was merged so can remove it

File: manim/mobject/vector_field.py (hunk 2 above)
Comment (ryanmccauley211) [reply manim/mobject/vector_field.py:385]: Added some docstrings on what I know about it, but I ported it from manimgl and besides some small tweaks I made I don't really know the details of it

File: manim/mobject/vector_field.py (hunk 4 above)
Comment (ryanmccauley211) [reply manim/mobject/vector_field.py]: Done, took the latter approach, as I think it is clearer than using a group

File: manim/mobject/types/opengl_vectorized_mobject.py (hunk 5 above)
Comment (ryanmccauley211) [reply manim/mobject/types/opengl_vectorized_mobject.py]: Removed the duplicate

**Review by hydrobeam — APPROVED** [review]
LGTM 🎉, Thank you for your efforts! I'll make a separate issue about the incompatibility with the projection shaders so that it can be handled later.

**Pull request merged**